components/script/serviceworkerjob.rs

```diff
@@ -176,105 +203,158 @@ impl JobQueue {
                 last_job.append_equivalent_job(job);
                 job_queue.push(last_job);
             } else {
+                // restore the popped last_job
                 job_queue.push(last_job);
+                // and push this new job to job queue
                 job_queue.push(job);
             }
         }
     }
 
     #[allow(unrooted_must_root)]
     // https://w3c.github.io/ServiceWorker/#run-job-algorithm
-    pub fn run_job(&mut self, run_job_handler: Box<RunJobHandler>, script_thread: &ScriptThread) {
+    pub fn run_job(&self, run_job_handler: Box<AsyncJobHandler>, script_thread: &ScriptThread) {
+        let queue_ref = &*self.0.borrow();
         let front_job =  {
-            let job_vec = self.0.get(&run_job_handler.scope_url);
+            let job_vec = queue_ref.get(&run_job_handler.scope_url);
             job_vec.unwrap().first().unwrap()
         };
-
         match front_job.job_type {
             JobType::Register => {
-                self.run_register(&front_job, run_job_handler, script_thread);
+                let handler = *run_job_handler;
+                let register_job_handler = AsyncJobHandler {
+                    global: handler.global,
+                    promise: handler.promise,
+                    scope_url: handler.scope_url,
+                    invoke_type: InvokeType::Register
+                };
+                script_thread.queue_register_job(box register_job_handler);
             },
             _ => {}
         }
     }
 
     #[allow(unrooted_must_root)]
     // https://w3c.github.io/ServiceWorker/#register-algorithm
-    pub fn run_register(&self, job: &Job, run_job_handler: Box<RunJobHandler>, script_thread: &ScriptThread) {
-        let global = run_job_handler.global.root();
-        let scope_url = run_job_handler.scope_url.clone();
+    pub fn run_register(&self, job: &Job, register_job_handler: Box<AsyncJobHandler>, script_thread: &ScriptThread) {
+        let global = register_job_handler.global.root();
+        // let job = (*job).first().unwrap();
+        let scope_url = register_job_handler.scope_url.clone();
```

Comment:
It's not clear to me why this is necessary, and if it is, then there are a lot of uses of `scope_url.clone()` in this method that are now unnecessary.